[19.0][MIG] sale_order_line_sequence: Migration to 19.0

BhaveshHeliconia opened the pull request.

nikul-serpentcs committed:
[12.0] [MIG] sale_order_line_sequence (#744)

* sale_order_line_sequence v10

* OCA Transbot updated translations from Transifex

* Update sale_order_line_sequence.pot

* Translated using Weblate (Portuguese)

Currently translated at 100,0% (6 of 6 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_order_line_sequence/pt/

* sale_order_line_sequence: migration to 11.0

* Update sale_order_line_sequence.pot

* [MIG] sale_order_line_sequence v12

oca-travis committed:
[UPD] Update sale_order_line_sequence.pot

oca-transbot committed:
Update translation files

Updated by Update PO files to match POT (msgmerge) hook in Weblate.

mvrodriguez committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_line_sequence/es/

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (8 of 8 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_line_sequence/de/

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (8 of 8 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_order_line_sequence/zh_CN/

kongrattapong committed:
[IMP] sale_order_line_sequense: black, isort

kongrattapong committed:
[MIG] sale_order_line_sequence: Migration to 13.0

rajpatel2240-odoo committed:
[MIG] sale_order_line_sequence: Migration to 15.0

MiquelRForgeFlow committed:
[UPD] Eficent -> ForgeFlow

MiquelRForgeFlow committed:
[MIG] sale_order_line_sequence: Migration to 14.0

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_order_line_sequence/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 25.0% (2 of 8 strings)

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_order_line_sequence/it/

OriolMForgeFlow committed:
[IMP] sale_order_line_sequence: module refactor, new sequence fields for orders and invoices

DorianMAG committed:
[IMP] sale_order_line_sequence: pre-commit stuff

DorianMAG committed:
add optional on sequence field in xml

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/es/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/es/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/it/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_order_line_sequence/pt_BR/

FSchiepers committed:
[MIG] sale_order_line_sequence: Migration to 17.0

Commit:
[UPD] Update sale_order_line_sequence.pot

antonioburic committed:
[IMP] linting changes to .rst files

sakthi-shan committed:
[IMP] sale_order_line_sequence: attrs removed

jhemono committed:
Translated using Weblate (French)

Currently translated at 70.0% (7 of 10 strings)

Translation: sale-workflow-17.0/sale-workflow-17.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-17-0/sale-workflow-17-0-sale_order_line_sequence/fr/

BT-mchechlacz committed:
[IMP] sale_order_line_sequence: pre-commit auto fixes

BT-mchechlacz committed:
[MIG] sale_order_line_sequence: Migration to 18.0

Co-authored-by: mchechlacz
Co-authored-by: Stefan Rijnhart <stefan@opener.amsterdam>

Commit:
[UPD] Update sale_order_line_sequence.pot

OCA-git-bot committed:
[BOT] post-merge updates

bo-eska committed:
Translated using Weblate (Turkish)

Currently translated at 100.0% (10 of 10 strings)

Translation: sale-workflow-18.0/sale-workflow-18.0-sale_order_line_sequence
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-18-0/sale-workflow-18-0-sale_order_line_sequence/tr/

BhaveshHeliconia committed:
[MIG] sale_order_line_sequence: Migration to 19.0

Quan-nhm reviewed (approved):
LGTM, code review and functional test.

alexey-pelykh commented on sale_order_line_sequence/models/account_move_line.py:
```diff
@@ -0,0 +1,24 @@
+                    f"{rec.sale_line_ids.visible_sequence}"
+                )
+            else:
+                rec.related_so_sequence = str(rec.sale_line_ids.visible_sequence)
```
Regression from 18.0. When `rec.sale_line_ids` contains multiple records, `rec.sale_line_ids.order_id.name` on a multi-record recordset will concatenate names (or raise, depending on Odoo version) rather than producing the expected `SO001/1, SO002/2` format.

Also missing the `len(rec.sale_line_ids) > 1` check that was in 18.0.

Please restore the 18.0 implementation:
```python
@api.depends("move_id.invoice_line_ids")
def _compute_related_so_sequence(self):
    for rec in self:
        if (
            len(rec.move_id.line_ids.sale_line_ids.order_id) > 1
            or len(rec.sale_line_ids) > 1
        ):
            sequences = sorted(
                set(
                    (sl.visible_sequence, sl.order_id.name)
                    for sl in rec.sale_line_ids
                )
            )
            rec.related_so_sequence = ", ".join(
                f"{seq[1]}/{seq[0]}" for seq in sequences
            )
        else:
            rec.related_so_sequence = str(
                rec.sale_line_ids.visible_sequence
            )
```

alexey-pelykh commented on sale_order_line_sequence/tests/test_sale_order_line_sequence.py:
```diff
@@ -0,0 +1,116 @@
+
+    def _create_sale_order(self):
+        sale_type = self.env["sale.order.type"].search([], limit=1)
+        vals = {
```
`sale.order.type` comes from `sale_order_type`, which is not in this module's `depends`. This will break on databases without that module.

The 18.0 tests did not reference this model. Please remove the `sale_type` lookup (line 37) and the `type_id` key (line 40).

alexey-pelykh commented on sale_order_line_sequence/tests/test_sale_order_line_sequence.py:
```diff
@@ -0,0 +1,116 @@
+        self.assertEqual(
+            str(so.order_line[1].visible_sequence),
+            self.invoice.line_ids[1].related_so_sequence,
+        )
```
The `test_combined_invoice_sequence` test from 18.0 was dropped. That test validates the multi-SO / merged-invoice-line scenario -- exactly the case broken by the `_compute_related_so_sequence` regression. Please re-add it, adapting for 19.0 conventions (`Command.create`, `product_uom_id`, `BaseCommon`).

alexey-pelykh reviewed (changes requested):
Thanks for the migration work. I found two issues that need to be addressed before this can be merged.

### 1. Regression in `_compute_related_so_sequence`

The 18.0 version correctly handles invoice lines linked to **multiple sale order lines** (e.g. merged invoice lines across SOs). It iterates over `rec.sale_line_ids`, builds sorted `(visible_sequence, order_name)` tuples, and joins them.

The 19.0 version simplified this and accesses `rec.sale_line_ids.order_id.name` / `rec.sale_line_ids.visible_sequence` directly in the f-string. When `sale_line_ids` is a multi-record recordset, this will produce incorrect results (concatenated strings like `SO001,SO002/1,2` instead of `SO001/1, SO002/2`). The condition also dropped the `len(rec.sale_line_ids) > 1` check.

The corresponding `test_combined_invoice_sequence` test that validated this scenario in 18.0 was also dropped.

### 2. Test references undeclared `sale.order.type`

`_create_sale_order` uses `self.env["sale.order.type"].search([], limit=1)` and passes `type_id` to the SO vals. This model comes from `sale_order_type`, which is **not** in this module's `depends`. The 18.0 tests did not use this. This will raise `KeyError` on a database without `sale_order_type` installed.

CI passed because `sale_order_type` likely happens to be present in the test environment, but the test itself is not portable.

BhaveshHeliconia commented:
@alexey-pelykh Thanks for the detailed review and for pointing these issues out. I've addressed the points you raised. Please have another look when you get a chance.

alexey-pelykh reviewed (approved):
All three items from my previous review are resolved:

1. `_compute_related_so_sequence` regression: fixed — multi-SO/merged-invoice logic restored from 18.0.
2. `sale.order.type` test dependency: removed.
3. `test_combined_invoice_sequence`: re-added and passing.

Code and tests look correct. CI green.

Co-Reviewed-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

rousseldenis commented:
/ocabot migration sale_order_line_sequence

rousseldenis commented:
/ocabot merge nobump
